### PR TITLE
Connect stack to a service account

### PR DESCRIPTION
**Pull request by safoinme**
## Describe changes
This branch implements a command line that connects every stack component within a stack at the time of command run to a given service connector.

## Pre-requisites
Please ensure you have done the following:
- [ ] I have read the **CONTRIBUTING.md** document.
- [ ] If my change requires a change to docs, I have updated the documentation accordingly.
- [ ] If I have added an integration, I have updated the [integrations](https://docs.zenml.io/stacks-and-components/component-guide) table and the [corresponding website section](https://zenml.io/integrations).
- [ ] I have added tests to cover my changes.
- [ ] I have based my new branch on `develop` and the open PR is targeting `develop`. If your branch wasn't based on develop read [Contribution guide on rebasing branch to develop](https://github.com/zenml-io/zenml/blob/main/CONTRIBUTING.md#-pull-requests-rebase-your-branch-on-develop).

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Other (add details above)

### CODE REVIEW

File: src/zenml/cli/stack.py [hunk 1]
```diff
@@ -1636,3 +1636,89 @@ def destroy(
...
+
+    For this reason, the `connect` command does not save the state of the
+    connection to the stack component. Instead, the state is saved to the
+    service-connector resource. This means that the `connect` should be run
+    every time a stack component is added to a stack, if the same service-
+    connector is used for all stack components.
+
+    The advantage of this approach is that the service-connector resource
+    can be used to connect to multiple stack components in one go, without
+    having to run the `connect` command multiple times.
```
Comment (fa9r) [inline src/zenml/cli/stack.py:1703]: IMO most of this is not relevant to users. If I understand it right, you could summarize it like this:

```suggestion
    """Connect a service-connector to all components of a stack.

    Note that this only connects the service-connector to the current 
    components of the stack and not to the stack itself, which means that 
    you need to rerun the command after adding new components to the stack.
```

File: src/zenml/cli/stack.py [hunk 2]
```diff
@@ -1636,3 +1636,89 @@ def destroy(
...
+
+    Args:
+        stack_name_or_id: Name of the stack for which to register secrets.
```
Comment (fa9r) [inline src/zenml/cli/stack.py:1706]:
```suggestion
        stack_name_or_id: Name of the stack to connect.
```

**Review by fa9r — APPROVED** [review]
Sure nice helper function I guess 👍

File: src/zenml/cli/stack.py [hunk 3]
```diff
@@ -1636,3 +1636,89 @@ def destroy(
...
+        interactive: Configure a service connector resource interactively.
+        no_verify: Skip verification of the connector resource.
+    """
+    from zenml.cli.stack_components import (
+        connect_stack_component_with_service_connector,
+    )
+
+    client = Client()
+    stack_to_connect = client.get_stack(name_id_or_prefix=stack_name_or_id)
+    for component in stack_to_connect.components.values():
```
Comment (strickvl) [inline src/zenml/cli/stack.py:1717]: What happens if this fails halfway through for some reason? If someone runs the command again will it just skip the ones that have already been connected?

Comment (safoinme) [reply src/zenml/cli/stack.py:1717]: it will go try to connect them again but they will be already connected

**Review by strickvl — APPROVED** [review]

File: src/zenml/cli/stack.py [hunk 4]
```diff
@@ -1636,3 +1636,76 @@ def destroy(
                 f"Spec directory for stack '{stack_name}' successfully deleted."
             )
         cli_utils.declare(f"Stack '{stack_name}' successfully destroyed.")
+
+
+@stack.command(
+    "connect", help="Connect a service-connector to a stack components."
```
Comment (stefannica) [inline src/zenml/cli/stack.py:1642]:
```suggestion
    "connect", help="Connect a service-connector to a stack's components."
```

File: src/zenml/cli/stack.py [hunk 5]
```diff
@@ -1636,3 +1636,76 @@ def destroy(
...
+    connector: Optional[str] = None,
+    interactive: bool = False,
+    no_verify: bool = False,
+) -> None:
+    """Connect a service-connector to all components of a stack.
+
+    Note that this only connects the service-connector to the current 
+    components of the stack and not to the stack itself, which means that 
+    you need to rerun the command after adding new components to the stack.
+
```
Comment (stefannica) [inline src/zenml/cli/stack.py:1691]: Could you store this note in the command's help text instead please ?

File: src/zenml/cli/stack.py [hunk 6]
```diff
@@ -1636,3 +1636,76 @@ def destroy(
...
+@click.option(
+    "--resource-id",
+    "-r",
+    "resource_id",
+    help="The resource ID to use with the connector. Only required for "
+    "multi-instance connectors that are not already configured with a "
+    "particular resource ID.",
+    required=False,
+    type=str,
+)
```
Comment (stefannica) [inline src/zenml/cli/stack.py:1662]:
```suggestion
```

This option is not used.

**Review by stefannica — CHANGES_REQUESTED** [review]

**Pull request merged**